upgrade-takeover.sh

```diff
@@ -0,0 +1,460 @@
+cat "${outfifo}" >> "${LOGFILE}" &
+# Read from the stderr FIFO, append to LOGFILE, and also display to terminal's stderr
+cat "${errfifo}" >> "${LOGFILE}" &
+# Redirect script's stdout and stderr to the respective FIFOs
+exec >"${outfifo}" 2>"${errfifo}"
+
+# Parse arguments
+while [[ $# -gt 0 ]]; do
+    arg="$1"
+
+    case $arg in
+        -h|--help)
+            help
+            exit 0
+            ;;
+        --force-slug)
+            if [ -z "$2" ]; then
+                log ERROR "\"$1\" argument needs a value."
+            fi
+            FORCED_SLUG=$2
+            shift
+            ;;
+        --hostos-version)
+            if [ -z "$2" ]; then
+                log ERROR "\"$1\" argument needs a value."
+            fi
+            target_version=$2
+            log "Raw target version: ${target_version}"
+            case $target_version in
+                *.prod)
+                    target_version="${target_version%%.prod}"
+                    log "Normalized target version: ${target_version}"
+                    ;;
+            esac
+            shift
+            ;;
+        --balenaos-registry)
+            if [ -z "$2" ]; then
+                log ERROR "\"$1\" argument needs a value."
+            fi
+            shift
+            ;;
+        *)
+            log WARN "Unrecognized option $1."
+            ;;
+    esac
+    shift
+done
+
+# Run on start
+_prepare_locking
+# Try to get the lock. Exit if unable; another instance is running already.
+exlock_now || exit 9
+
+if [ -z "$target_version" ]; then
+    log ERROR "--hostos-version is required."
+fi
+
+progress 25 "Preparing OS update"
+
+# Retrieve slug (device type) from API, and use this value if not provided to  the script.
+# Verify this call succeeds?
+FETCHED_SLUG=$(CURL_CA_BUNDLE="${TMPCRT}" ${CURL} -H "Authorization: Bearer ${APIKEY}" \
+    "${API_ENDPOINT}/v6/device?\$select=is_of__device_type&\$expand=is_of__device_type(\$select=slug)&\$filter=uuid%20eq%20%27${UUID}%27" 2>/dev/null \
+    | jq -r '.d[0].is_of__device_type[0].slug'
+    )
+
+SLUG=${FORCED_SLUG:-$FETCHED_SLUG}
+
+# Validate target version in semver (major > 1) or year.month.patch format
+if [ -n "$target_version" ]; then
+    case $target_version in
+        [2-9].*|[1-9][0-9].*|2[0-9][0-9][0-9].*.*)
+            log "Target OS version \"$target_version\" OK."
+            ;;
+        *)
+            log ERROR "Target OS version \"$target_version\" not supported."
+            ;;
+    esac
+else
+    log ERROR "No target OS version specified."
+fi
+
+# Validate host OS version, similar to target_version above
+case $VERSION in
+    [2-9].*|[1-9][0-9].*|2[0-9][0-9][0-9].*.*)
+        log "Host OS version \"$VERSION\" OK."
+        ;;
+    *)
+        log ERROR "Host OS version \"$VERSION\" not supported."
+        ;;
+esac
+
+# Ensure takeover environment is prepared
+takeover_path="/mnt/data/takeover"
+mkdir -p "${takeover_path}"
+
+# Download target hostOS image as required
```

Comment:
Is it feasible to avoid re-downloading the same OS image in case of running again after failure?